drivers/gles3/shaders/scene.glsl

```diff
@@ -1059,12 +1059,17 @@ void main() {
 	normal_map.z = sqrt(max(0.0, 1.0 - dot(normal_map.xy, normal_map.xy))); //always ignore Z, as it can be RG packed, Z may be pos/neg, etc.
 
 	normal = normalize(mix(normal, tangent * normal_map.x + binormal * normal_map.y + normal * normal_map.z, normal_map_depth));
-
+#elif defined(NORMAL_USED)
+	normal = normalize(normal);
 #endif
 
 #ifdef LIGHT_ANISOTROPY_USED
+	// Reconstructing orthonormal tangent frame from normal map result means that anisotropic specularity can be synced to the tangent basis.
+	// Without this, anisotropy can look wibbly-wobbly in extreme cases and dark around the edges otherwise.
+	tangent = normalize(cross(binormal, normal));
+	binormal = normalize(cross(normal, tangent));
```

Comment:
I am pretty sure you are totally right. Great catch.

Reply:
To answer your questions:
1. ``TANGENT_USED`` is defined if the user accesses the ``TANGENT`` built-in in their shader. It ensures that the tangent attribute is available
2. ``LIGHT_ANISOTROPY_USED`` also ensures that the tangent attribute is available. 

My concern with recreating the tangent space is that it adds a significant cost to all shaders using anisotropy to solve an issue that only occurs when using normal mapping combined with anisotropy

> There is a possibility that if tangent space is used in user shaders they will need to reconstruct and normalize it if anisotropy isn't enabled.

Is this with this PR, or in master?

Reply:
It's not as much an issue in master because in master vec3 tangent = normalize(tangent_interp) is done before #CODE : FRAGMENT which is where I assume user shaders are injected. If a user shader assumes the tangent space will be normalized (which would lead to incorrect but not obviously incorrect results when applying normal maps in a user shader) it would be broken especially by scaling down after this PR.

I'm not too familiar with what types of things are done in user shaders so I'm not sure how big of an issue it will be to do things in a more correct way.

Anisotropy is generally applied to a relatively small portion of the scene. The tangent space reconstruction consists of normalization, which is necessary anyway and is already done per fragment in master, and 1 additional cross product for each of the tangent and binormal, which should be relatively cheap compared to the normalization. It's also necessary for correct results of the anisotropic specular highlight later on. You can see a difference in the position of the specular highlight in master in the material preview when toggling anisotropy on and off even when there's no normal map applied. That's because we aren't reconstructing the tangent space after calculating the normal. It's incorrect _even when there isn't a normal map applied_ and possibly even more incorrect when there is. That's why I think we should reconstruct the tangent basis when anisotropy is enabled. It might be possible to do it more efficiently by applying a quaternion to both vectors, but I think it's a good idea to do it in any case.